pkg/bootkube/create.go

```diff
@@ -43,143 +59,133 @@ func CreateAssets(config clientcmd.ClientConfig, manifestDir string, timeout tim
-		return err
+	// System namespace has been created
+	_, err = client.CoreV1().Namespaces().Get("kube-system", metav1.GetOptions{})
+	return err
+}
+
+type manifest struct {
+	kind       string
+	apiVersion string
+	namespace  string
+	name       string
+	raw        []byte
+
+	filepath string
+}
+
+func (m manifest) String() string {
+	if m.namespace == "" {
+		return fmt.Sprintf("%s %s %s", m.filepath, m.kind, m.name)
 	}
+	return fmt.Sprintf("%s %s %s/%s", m.filepath, m.kind, m.namespace, m.name)
+}
 
-	filenameOpts := &resource.FilenameOptions{
-		Filenames: []string{manifestDir},
-		Recursive: false,
+type creater struct {
+	client *rest.RESTClient
+
+	// mapper maps resource kinds ("ConfigMap") with their pluralized URL
+	// path ("configmaps") using the discovery APIs.
+	mapper *resourceMapper
+}
+
+func newCreater(c *rest.Config) (*creater, error) {
+	c.NegotiatedSerializer = serializer.DirectCodecFactory{CodecFactory: scheme.Codecs}
+	client, err := rest.UnversionedRESTClientFor(c)
+	if err != nil {
+		return nil, err
 	}
 
-	r := builder.
-		Schema(schema).
-		ContinueOnError().
-		NamespaceParam(cmdNamespace).DefaultNamespace().
-		FilenameParam(enforceNamespace, filenameOpts).
-		SelectorParam("").
-		Flatten().
-		Do()
-	err = r.Err()
+	discoveryClient, err := discovery.NewDiscoveryClientForConfig(c)
 	if err != nil {
-		return err
+		return nil, err
 	}
 
-	count := 0
-	err = r.Visit(func(info *resource.Info, err error) error {
-		if err != nil {
-			return err
-		}
+	return &creater{
+		mapper: newResourceMapper(discoveryClient),
+		client: client,
+	}, nil
+}
 
-		obj, err := resource.NewHelper(info.Client, info.Mapping).Create(info.Namespace, true, info.Object)
-		if err != nil {
-			if apierrors.IsAlreadyExists(err) {
-				count++
-				return nil
-			}
-			return cmdutil.AddSourceToErr("creating", info.Source, err)
+func (c *creater) createManifests(manifests []manifest) error {
+	var crds, other []manifest
+	for _, m := range manifests {
+		if m.kind == "CustomResourceDefinition" && strings.HasPrefix(m.apiVersion, "apiextensions.k8s.io/") {
+			crds = append(crds, m)
+		} else {
+			other = append(other, m)
 		}
-		info.Refresh(obj, true)
-		gvk := obj.GetObjectKind().GroupVersionKind()
+	}
 
-		// If the object is a CRD, wait for it to fully roll out before continuing.
-		// This is because we may also be creating other instances of this CRD later.
-		if gvk.Kind == "CustomResourceDefinition" {
-			if err = waitForCRDRollout(info.Client, gvk, info.Object); err != nil {
-				return err
-			}
+	// Create the custom resource definition before creating the actual custom resources.
+	for _, m := range crds {
+		if err := c.create(m); err != nil {
+			return fmt.Errorf("creating %s: %v", m, err)
 		}
+		UserOutput("Created %s\n", m)
+	}
 
-		count++
-		shortOutput := false
-		if !shortOutput {
-			f.PrintObjectSpecificMessage(info.Object, util.GlogWriter{})
+	for _, crd := range crds {
+		if err := c.waitForCRD(crd); err != nil {
```

Comment:
What is `waitForCRD` doing?

Reply:
It got brought over from the old code. It's just waiting for the API server to respond to a "list" on the custom resource. Will add a comment to the function.